vnext/JSI/Universal/ChakraRtRuntime.cpp

```diff
@@ -0,0 +1,121 @@
+      std::static_pointer_cast<const ChakraPreparedJavaScript>(preparedJs);
+
+  JsValueRef result = nullptr;
+
+  std::wstring sourceCodeUtf16 = Common::Unicode::Utf8ToUtf16(
+      Common::Utilities::CheckedReinterpretCast<const char *>(chakraPreparedJs->SourceCode()->data()),
+      chakraPreparedJs->SourceCode()->size());
+
+  std::wstring sourceUrlUtf16 = Common::Unicode::Utf8ToUtf16(chakraPreparedJs->SourceUrl());
+
+  // We have to use a const_cast here due to JsRunSerializedScript taking in a
+  // BYTE* instead of a const BYTE*.
+  JsErrorCode error = JsRunSerializedScript(
+      sourceCodeUtf16.c_str(),
+      static_cast<BYTE *>(const_cast<uint8_t *>(chakraPreparedJs->Bytecode()->data())),
+      JS_SOURCE_CONTEXT_NONE /*sourceContext*/,
+      sourceUrlUtf16.c_str(),
+      &result);
+
+  // Fall back to evaluating source code if bytecode evaluation fails.
+  if (error == JsErrorBadSerializedScript) {
+    JsValueRef exception = nullptr;
+    VerifyJsErrorElseThrow(JsGetAndClearException(&exception));
+    return EvaluateJavaScriptImpl(sourceCodeUtf16.c_str(), sourceUrlUtf16.c_str());
+  }
+
+  VerifyJsErrorElseThrow(error);
+  return ToJsiValue(ChakraObjectRef(result));
+}
+
+std::string ChakraRtRuntime::description() {
+  return "ChakraRtRuntime";
+}
+
+facebook::jsi::WeakObject ChakraRtRuntime::createWeakObject(const facebook::jsi::Object &object) {
+  return make<facebook::jsi::WeakObject>(CloneChakraPointerValue(getPointerValue(object)));
+}
+
+facebook::jsi::Value ChakraRtRuntime::lockWeakObject(const facebook::jsi::WeakObject &weakObject) {
+  // We need to make a copy of the ChakraObjectRef held within weakObj's
+  // member PointerValue for the returned jsi::Value here.
+  ChakraObjectRef ref = GetChakraObjectRef(weakObject);
+  return ToJsiValue(std::move(ref));
+}
+
+facebook::jsi::Value ChakraRtRuntime::EvaluateJavaScriptSimple(
+    const std::shared_ptr<const facebook::jsi::Buffer> &buffer,
+    const std::string &sourceUrl) {
+  assert(buffer);
+
+  const std::wstring sourceCodeUtf16 = Common::Unicode::Utf8ToUtf16(
+      Common::Utilities::CheckedReinterpretCast<const char *>(buffer->data()), buffer->size());
+
+  const std::wstring sourceUrlUtf16 = Common::Unicode::Utf8ToUtf16(sourceUrl);
+
+  return EvaluateJavaScriptImpl(sourceCodeUtf16.c_str(), sourceUrlUtf16.c_str());
+}
+
+std::shared_ptr<const facebook::jsi::Buffer> ChakraRtRuntime::SerializeScript(
+    const std::shared_ptr<const facebook::jsi::Buffer> &buffer) {
+  assert(buffer);
+
+  std::wstring sourceCodeUtf16 = Common::Unicode::Utf8ToUtf16(
+      Common::Utilities::CheckedReinterpretCast<const char *>(buffer->data()), buffer->size());
+
+  unsigned long bytecodeSize = 0;
+  VerifyJsErrorElseThrow(JsSerializeScript(sourceCodeUtf16.c_str(), nullptr, &bytecodeSize));
+
+  assert(
+      static_cast<unsigned long long>(bytecodeSize) <=
+      static_cast<unsigned long long>((std::numeric_limits<size_t>::max)()));
```

Comment:
There are plenty of places where we are checking for overflow/underflow before an integer conversion. It would be nice to implement and use something like boost::numeric_cast. In my opinion it would be nice to just use boost::numeric_cast instead of implementing our own. Thoughts?

Reply:
We already have a dependency on boost. I would just use their version.

---
In reply to: [345396549](https://github.com/microsoft/react-native-windows/pull/3650#discussion_r345396549) [](ancestors = 345396549)